File: api/src/main/java/com/spontancombust/workoutoclock/services/AuthService.java
package com.spontancombust.workoutoclock.services;

import java.time.Instant;
import java.util.Date;

import org.springframework.security.authentication.AuthenticationCredentialsNotFoundException;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import lombok.AllArgsConstructor;

import com.spontancombust.workoutoclock.exceptions.EmailTakenException;
import com.spontancombust.workoutoclock.model.RefreshToken;
import com.spontancombust.workoutoclock.model.AuthTokenPair;
import com.spontancombust.workoutoclock.model.User;
import com.spontancombust.workoutoclock.repositories.UserRepository;
import com.spontancombust.workoutoclock.security.JwtService;
import com.spontancombust.workoutoclock.security.RefreshTokenService;
import com.spontancombust.workoutoclock.security.UserPrincipal;
import com.spontancombust.workoutoclock.security.UserPrincipalService;


public interface AuthService {
    User signUp(String email, String password, String username) throws EmailTakenException;

    //TODO rate-limiting for incorrect credentials
    AuthTokenPair signIn(String email, String password) throws AuthenticationException;

    AuthTokenPair refreshAuth(String refreshToken);

    void signOut(String refreshToken);

    //TODO change password


    User getSignedInUser();
}


@Service
@AllArgsConstructor
class AuthServiceImpl implements AuthService {

    private final JwtService jwtService;
    private final RefreshTokenService refreshTokenService;
    private final UserPrincipalService userPrincipalService;

    private final AuthenticationManager authenticationManager;
    private final PasswordEncoder passwordEncoder;
    private final UserRepository userRepository;


    @Override
    public AuthTokenPair signIn(String email, String password) throws AuthenticationException {
        var auth = this.authenticationManager.authenticate(
            new UsernamePasswordAuthenticationToken(email, password)
        );

        SecurityContextHolder.getContext().setAuthentication(auth);
        var principal = (UserPrincipal)auth.getPrincipal();

        String accessToken = this.jwtService.issueToken(principal);
        RefreshToken refreshToken = this.refreshTokenService.issueRefreshToken(principal.getUserId());

        return new AuthTokenPair(accessToken, refreshToken.getTokenString());
    }

    @Override
    public User signUp(String email, String password, String username) throws EmailTakenException {
        if (this.userRepository.existsByEmail(email)) {
            throw new EmailTakenException();
        }

        User newUser = User.builder()
                        .email(email)
                        .password(this.passwordEncoder.encode(password))
                        .username(username)
                        .createdDate(Date.from(Instant.now()))
                        .build();

        //TODO email confirmation

        User savedUser = this.userRepository.save(newUser);
        return savedUser;
    }

    @Override
    public AuthTokenPair refreshAuth(String refreshToken) {
        RefreshToken newRefreshToken = this.refreshTokenService.rotateRefreshToken(refreshToken);

        UserPrincipal principal = this.userPrincipalService.loadUserByModel(newRefreshToken.getUser());
        String newAccessToken = this.jwtService.issueToken(principal);

        return new AuthTokenPair(newAccessToken, newRefreshToken.getTokenString());
    }

    @Override
    public void signOut(String refreshToken) {
        this.refreshTokenService.invalidateRefreshToken(refreshToken);
    }


    @Override
    public User getSignedInUser() throws AuthenticationException {
        var auth = SecurityContextHolder.getContext().getAuthentication();

        if (auth == null) {
            throw new AuthenticationCredentialsNotFoundException("User not logged in");
        }

        var principal = (UserPrincipal) auth.getPrincipal();
        var user = this.userRepository.findById(principal.getUserId()).get();

        return user;
    }
}

File: client/src/services/AuthService.ts
import { ChangePasswordRequestDto, RefreshRequestDto, SignInRequestDto, SignOutRequestDto, SignUpRequestDto } from "../api/dto/AuthDtos";
import Api from "../api/requests";
import Converters from "../converters/Converters";
import AuthTokenPair from "../types/AuthTokenPair";
import User from "../types/User";


export default class AuthService {

    async signUp(email: string, password: string, username: string) : Promise<User> {
        const req: SignUpRequestDto = {
            email,
            password,
            username
        };

        const respDto = await Api.auth().signUp(req);

        return Converters.user.fromDto(respDto);
    }

    async signIn(email: string, password: string) : Promise<AuthTokenPair> {
        const req: SignInRequestDto = {
            email,
            password
        };

        const respDto = await Api.auth().signIn(req);

        return {
            accessToken: respDto.accessToken,
            refreshToken: respDto.refreshToken
        }
    }

    async refresh(refreshToken: string) : Promise<AuthTokenPair> {
        const req: RefreshRequestDto = {
            refreshToken
        };

        const respDto = await Api.auth().refresh(req);

        return {
            accessToken: respDto.accessToken,
            refreshToken: respDto.refreshToken
        }
    }

    async signOut(refreshToken: string) {
        const req: SignOutRequestDto = {
            refreshToken
        };

        await Api.auth().signOut(req);
    }

    async changePassword(oldPassword: string, newPassword: string) : Promise<User> {
        const req: ChangePasswordRequestDto = {
            oldPassword,
            newPassword
        };

        const respDto = await Api.auth().changePassword(req);

        return Converters.user.fromDto(respDto);
    }


    async getSignedInUser() : Promise<User> {
        const respDto = await Api.auth().getSignedInUser();

        return Converters.user.fromDto(respDto);
    }
}